onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -259,6 +259,11 @@ bool SetDynamicRange(nvinfer1::INetworkDefinition& network, std::unordered_map<s
   }
   return true;
 }
+
+int GetSubgraphKernelIndex() {
```

Comment:
on second thought, static may not work for the case where there are multiple model loads.
across process restarts, the counter value may not be consistent. and the inconsistent naming would result in engine caching to not work properly.

Reply:
If the models are different, most likely hash value in the engine name is also different. If the models are the same, the order the subgraph kernels (also the indexing) are generated is also the same, so I think the engine can also be reused.

Reply:
i'm talking about for the same model, the kernels index could be different depending on model load order. 
let's say you have model 1 with 2 subgraphs and model 2 with 1 subgraph and you create a session for each model.
if you load model 1 first,  its subgraph would have counter value 0 and 1.
however, if you load model 2 first, model 1's subgraph will have counter value 1 and 2.
and this would break engine caching, right?

Reply:
Different models most likely will have different subgraphs, therefore different engine names, so engine caching would not be broken, but regenerated.

Reply:
> Different models most likely will have different subgraphs, therefore different engine names, so engine caching would not be broken, but regenerated.

it's broken in the sense that it's a cache miss instead of a hit.  the idea is to not regenerate the engine because we already created the engine already previously.
because of that static/global counter, it keeps incrementing so the name could change. 
that's not the behavior we want. if it was already doing that previously, then we should fix it. 
we want the operation to be idempotent and deterministic no matter how many models are loaded.
let's discuss more offline if there's something i'm missing.

Reply:
Not sure if this is possible as I'm not familiar with how a TRT EP instance is used across multiple sessions. 

Are things simpler if we layer the ORT EP instance over the underlying TRT instance so that there is a 1:1 between the ORT session EP instance and a model load but the TRT instance has a 1:M with the EP instances? With that, you can be deterministic about the subgraph id and this sort of approach may suffice:

https://github.com/microsoft/onnxruntime/blob/d4dddd99d9e42d35247bbcbbb0c87feffb47f633/onnxruntime/test/providers/internal_testing/internal_testing_execution_provider.h#L29

Reply:
Had an offline discussion with @skottmckay. We agreed there are two requirements for the kernel naming,
1. For same model, the name should be deterministic in each session and unique for every TRT subgraph
2. Across different models, the name should be also unique but it can not be 100% guaranteed. we may hash as many meta data as possible to the name to reduce the collision.
Thank @skottmckay for drafting code for hashing parent graph info to the name, but in order to use it in the EP we need to change TRT shared library as well. Also we may hash other info such as ORT version, GPU info etc as we tried before. Given than I prefer to have hashing changes in a separate PR.